source/server/connection_handler_impl.cc

```diff
@@ -288,5 +310,60 @@ ListenerStats ConnectionHandlerImpl::generateStats(Stats::Scope& scope) {
   return {ALL_LISTENER_STATS(POOL_COUNTER(scope), POOL_GAUGE(scope), POOL_HISTOGRAM(scope))};
 }
 
+ConnectionHandlerImpl::ActiveUdpListener::ActiveUdpListener(ConnectionHandlerImpl& parent,
+                                                            Network::ListenerConfig& config)
+    : ActiveUdpListener(parent, parent.dispatcher_.createUdpListener(config.socket(), *this),
+                        config) {}
+
+ConnectionHandlerImpl::ActiveUdpListener::ActiveUdpListener(ConnectionHandlerImpl& parent,
+                                                            Network::ListenerPtr&& listener,
+                                                            Network::ListenerConfig& config)
+    : ConnectionHandlerImpl::ActiveListenerBase(parent, std::move(listener), config),
+      udp_listener_(dynamic_cast<Network::UdpListener*>(listener_.get())) {
+
+  RELEASE_ASSERT(udp_listener_ != nullptr, "");
+
+  // Create the filter chain on creating a new udp listener
+  config_.filterChainFactory().createUdpListenerFilterChain(*this);
+
+  // If filter chain is empty, fail the creation of the listener
+  if (read_filters_.empty()) {
+    throw Network::CreateListenerException(fmt::format(
+        "Cannot create listener as no read filters registered for the udp listener: {} ",
+        config_.name()));
+  }
+}
+
+void ConnectionHandlerImpl::ActiveUdpListener::onData(Network::UdpRecvData& data) {
+  for (auto& filter : read_filters_) {
```

Comment:
We can only have 1 filter now, right? Let's assert that and avoid the for loop. Same elsewhere.